web/src/message_view.ts

```diff
@@ -400,6 +401,23 @@ export let show = (raw_terms: NarrowTerm[], show_opts: ShowMessageViewOpts): voi
         raw_terms = [{operator: "in", operand: "home"}];
     }
     const filter = new Filter(raw_terms);
+
+    // If the filter is in channel topic narrow, but we don't have any of
+    // the messages of it locally, then we set `narrow_requires_hash_change`
+    // true.
+    // This would make sure that when messages are fetched from server, if
+    // the narrow has any message we update the hash to a topic permalink
+    // by attaching `with` operator to the last message.
+    if (filter.has_exactly_channel_topic_operators()) {
+        const channel_id = stream_data.get_stream_id(filter.operands("channel")[0]!)!;
+        const topic = filter.operands("topic")[0]!;
+        const last_msg_id = get_latest_known_message_id_in_topic(channel_id, topic);
```

Comment:
I don't think this is ideal when we have a `then_select_id`. What's the reasoning to not do this adjustment at the end, when we are selecting a specific message ID?